src/main/java/org/opentripplanner/routing/algorithm/transferoptimization/model/OptimizedPathTail.java

```diff
@@ -17,36 +25,132 @@
+    private int transferPriorityCost = TransferConstraint.ZERO_COST;
+    private int waitTimeOptimizedCost = TransferWaitTimeCalculator.ZERO_COST;
 
     public OptimizedPathTail(
-            TransitPathLeg<T> leg,
-            Map<PathLeg<T>, ConstrainedTransfer> transfersTo,
-            int transferPriorityCost,
-            int waitTimeOptimizedCost
+            RaptorSlackProvider slackProvider,
+            CostCalculator costCalculator,
+            TransferWaitTimeCalculator waitTimeCostCalculator
+            ) {
+        super(null, slackProvider, costCalculator);
+        this.waitTimeCostCalculator = waitTimeCostCalculator;
+    }
+
+    private OptimizedPathTail(OptimizedPathTail<T> other) {
+        super(other);
+        this.waitTimeCostCalculator = other.waitTimeCostCalculator;
+        this.waitTimeOptimizedCost = other.waitTimeOptimizedCost;
+        this.transferPriorityCost = other.transferPriorityCost;
+    }
+
+    @Override
+    protected void add(PathBuilderLeg<T> newLeg) {
+        addHead(newLeg);
+        // Keep from- and to- times up to date by time-shifting access, transfer and egress legs.
+        newLeg.timeShiftThisAndNextLeg(slackProvider);
+        addTransferPriorityCost(newLeg);
+        addOptimizedWaitTimeCost(newLeg);
+    }
+
+    @Override
+    protected void updateAggregatedFields() {
+        /* Empty, aggregated fields are updated while adding new legs */
+    }
+
+    /**
+     * Create a deep-copy of this builder.
+     */
+    public OptimizedPathTail<T> mutate() {
+        return new OptimizedPathTail<>(this);
+    }
+
+    /** Start by adding the last transit leg with the egress leg attached. */
+    public OptimizedPathTail<T> addTransitTail(TransitPathLeg<T> leg) {
+        egress(leg.nextLeg().asEgressLeg().egress());
+        var times = new BoardAndAlightTime(leg.trip(), leg.getFromStopPosition(), leg.getToStopPosition());
+        transit(leg.trip(), times);
+        return this;
+    }
+
+    /**
+     * Insert a new transit leg at the head and return the new object. The new tail is returned
+     * with the given transit + transfer leg, earliest-departure-time and the current leg as a
+     * new tail.
+     */
+    public OptimizedPathTail<T> addTransitAndTransferLeg(
+            TransitPathLeg<T> originalLeg,
+            TripToTripTransfer<T> tx
+    ) {
+        head().changeBoardingPosition(tx.to().stopPosition());
+
+        if(!tx.sameStop()) {
+            transfer(tx.getPathTransfer(), tx.to().stop());
+        }
+
+        // The transfer may happen before the original board point. If so, the boarding must be
+        // changed so the leg is valid (not traveling in reverse/back in time). Also, setting
+        // the board position to the first stop in the pattern make sure all paths starts at the
+        // same place; hence the generalized-cost can be compared.
+        // The board position will be changed when a new head is inserted.
```

Comment:
```suggestion
        // The transfer may happen before the original boarding point. If so, the boarding must be
        // changed so that the leg is valid (not traveling in reverse/back in time). Also, setting
        // the boarding position to the first stop in the pattern makes sure that all paths start at the
        // same place; hence the generalized-cost can be compared.
        // The board position will be changed when a new head is inserted.
```